{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "dcc5c21a",
   "metadata": {},
   "source": [
    "# Enrichment Score for specific k-mer \n",
    "\n",
    "Input: \n",
    "1. binsize \n",
    "2. specific k-mer\n",
    "3. Fragment Annotation and chromosome size file\n",
    "\n",
    "Output:\n",
    "1. 100 genomewide permutated randomly selected k-mer counts\n",
    "2. summary of k-mer count "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0acf4dee",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import datatable as dt\n",
    "import random\n",
    "import gc\n",
    "import os,time\n",
    "import math"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "228047bb",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Loading\n",
    "def LoadMatrixReader(filename, Chunksize, usecols,  sepstr=\"\\t\"):\n",
    "    file_reader = pd.read_table(filename, sep=sepstr,chunksize=Chunksize, iterator=True, \n",
    "                                index_col=None, usecols = usecols)\n",
    "    return(file_reader)\n",
    "\n",
    "\n",
    "# Binsize\n",
    "def Binsize(binsize, p_array):\n",
    "    p_array = np.divide(p_array, binsize).astype(\"int\")*binsize\n",
    "    return(p_array)\n",
    "\n",
    "# Remove the fragments with the same bins\n",
    "def Deduplicate(df, binsize):\n",
    "    df.loc[:, \"posbin\"] = Binsize(binsize, (df[\"start\"].values + df[\"end\"].values)/2 )\n",
    "    # fragmetn bin相同的时候，去其中一个\n",
    "    Dedup_df = df.groupby([\"read_name\",\"posbin\"], as_index=False).first()\n",
    "    Dedup_df = Dedup_df.reset_index()\n",
    "    Dedup_df = Dedup_df.loc[:, [\"read_name\",\"chrom\", \"posbin\"] ]\n",
    "    return(Dedup_df)\n",
    " \n",
    "def RandomKmers(inputBins, randombins, randomchroms):\n",
    "    '''\n",
    "    Calculate Kmer for randomselect bins\n",
    "    '''\n",
    "    startbin = min(inputBins)\n",
    "    Kmerdist = [ pb-startbin for pb in inputBins ] # input kermer distance\n",
    "    \n",
    "    randKmer = {\"chrom\":[], \"Kmer\":[] }\n",
    "    RandomkmerN = 0\n",
    "    for chrom, posbin in zip(randomchroms, randombins):\n",
    "        Kmer = [posbin+kdist for kdist in  Kmerdist ]\n",
    "        if max(Kmer) <=  Genenomelen[chrom]:\n",
    "            randKmer[\"chrom\"].append(chrom)\n",
    "            randKmer[\"Kmer\"].append(Kmer)\n",
    "            RandomkmerN  += 1\n",
    "    print(\"Export %d Random Kmers\"%(RandomkmerN) )\n",
    "    return(randKmer)    \n",
    "\n",
    "def FragmentFilter(df, ObsKmer):\n",
    "    '''\n",
    "    Filter Fragments with random select bins\n",
    "    '''\n",
    "    # 1  Filter select bins\n",
    "    P = df[\"chrom\"].str.cat(df[\"posbin\"].astype(str) ,sep=\"_\").isin(Allchrbins) \n",
    "    df = df.loc[P, :]\n",
    "    # 2 Filter reads with fragment numbers larger than Kmer\n",
    "    Kmer_frags = df.groupby(\"read_name\")[\"chrom\"].count()\n",
    "    Pnum = Kmer_frags >= len(ObsKmer)\n",
    "    df = df.loc[ df.read_name.isin(Kmer_frags[Pnum].index), : ]\n",
    "    df = df .sort_values(by=[\"chrom\",\"read_name\", \"posbin\" ], ignore_index=True)\n",
    "    return(df)\n",
    "\n",
    "def CountKmers(df, ObsKmer):\n",
    "    '''\n",
    "    '''\n",
    "    Kmerbins = len(ObsKmer)\n",
    "    for (read_name, chrom), gdf in df.groupby([\"read_name\", \"chrom\"]):\n",
    "        posbins = gdf.posbin.to_list()\n",
    "        if len(posbins) == Kmerbins:\n",
    "            posbins = sorted(posbins)\n",
    "            kmerkey = \"%s_%s\"%(chrom, \"_\".join([str(pb) for pb in posbins])   )\n",
    "            try:\n",
    "                RandomKmerCount[kmerkey] += 1 # Random Kmer count, Kmer match\n",
    "            except:\n",
    "                pass\n",
    "    return"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d6a6cd2e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Kmers\n",
    "binsize = 25000\n",
    "ObsKmerChrom, ObsKmer, Obscount =\"chr8\", [132850000, 133200000, 133450000], 26\n",
    "\n",
    "\n",
    "# Random Select Kmers\n",
    "## genome position bins\n",
    "chromlen_file = \"/data1/ZJY/Pore-C/HiGlass/data/hg38.chromosomesize.csv\"\n",
    "chromlen_df = pd.read_csv(chromlen_file, sep=\"\\t\", header=None, index_col=None, names=[\"chrom\", \"len\"])\n",
    "Genenomeposbin = {\"chrom\":[],\n",
    "              \"posbin\":[] }\n",
    "Genenomelen = {}\n",
    "for n, rowvalue in chromlen_df.iterrows():\n",
    "    chrom = rowvalue[\"chrom\"]\n",
    "    if chrom == \"chrY\":\n",
    "        continue\n",
    "    s_pbs = binsize \n",
    "    posbins =  list( np.arange(10*binsize, rowvalue[\"len\"]-10*binsize, binsize) ) # all position bins, except the telomere\n",
    "    chroms = len(posbins) * [chrom]\n",
    "    Genenomeposbin [\"chrom\"].extend( chroms )\n",
    "    Genenomeposbin [\"posbin\"].extend( posbins )\n",
    "    # Genome len \n",
    "    Genenomelen[chrom] = rowvalue[\"len\"]\n",
    "Genenomeposbin  = pd.DataFrame(Genenomeposbin )\n",
    "\n",
    "# Random select bins\n",
    "SelectNum = 100\n",
    "\n",
    "rindex = random.sample(Genenomeposbin.index.to_list(), SelectNum)\n",
    "randKmer = RandomKmers(ObsKmer, Genenomeposbin.loc[rindex, \"posbin\"].to_list() , Genenomeposbin.loc[rindex, \"chrom\"].to_list() )\n",
    "\n",
    "## random chrombins and counts\n",
    "Allchrbins = []\n",
    "RandomKmerCount = {}\n",
    "for chrom, posbins in zip( randKmer[\"chrom\"], randKmer[\"Kmer\"] ):\n",
    "    chrbin = [\"%s_%d\"%(chrom, pb) for pb in posbins ]\n",
    "    Allchrbins.extend(chrbin)\n",
    "    \n",
    "    kmerkey = \"%s_%s\"%(chrom, \"_\".join([str(pb) for pb in posbins])   )\n",
    "    RandomKmerCount[kmerkey] = 0 # Random Kmer count"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8d87f554",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Loading\n",
    "filename = \"/data1/ZJY/Pore-C/Analysis/NA12878/Merge_NA12878/vdFAnnotation/Merge_Align_Fragment_RvdF_5Reps.csv\"\n",
    "Chunksize = 10000000\n",
    "\n",
    "usecols = [\"read_name\", \"chrom\", \"start\", \"end\"]\n",
    "df_reader = LoadMatrixReader(filename, Chunksize,  usecols, \",\")\n",
    "\n",
    "for df in df_reader:\n",
    "    print(\"Loading %d records\"%len(df) )\n",
    "    df = Deduplicate(df, binsize)\n",
    "    df = FragmentFilter(df, ObsKmer)\n",
    "    if len(df) >= 1:\n",
    "        CountKmers(df, ObsKmer)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e3f0eb5a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Export\n",
    "Count_df = pd.DataFrame({\"Random_Kmers\":RandomKmerCount.keys(),\n",
    "                         \"Count\":RandomKmerCount.values()})\n",
    "Exportdir = \"/data1/ZJY/Pore-C/Analysis/Jupyter/Figures/Cluster/Fig6_7_0410/EnrichScore\"\n",
    "os.system('mkdir -p %s'%Exportdir)\n",
    "kmercount_file = \"%s/%s_%s_permutated_kmers_count.csv\"%(Exportdir, ObsKmerChrom, \"_\".join([str(pb) for pb in ObsKmer]) )\n",
    "Count_df.to_csv(kmercount_file, header=True, index=False, sep=\"\\t\" )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "857056c6",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Calculate  OE\n",
    "summary_result = []\n",
    "ExpertCount = Count_df[\"Count\"].mean()\n",
    "print(Count_df[\"Count\"].describe() )\n",
    "print(\"Expected K-mer counts:%.3f\"%ExpertCount)\n",
    "OE = Obscount/ExpertCount\n",
    "print(\"O/E=%.3f\"%OE)\n",
    "n\n",
    "summary_result = []\n",
    "summary_result.append( str(Count_df[\"Count\"].describe())  )\n",
    "summary_result.append( \"Observed_K-mer_Count=%d\"%Obscount  )\n",
    "summary_result.append( \"O/E=%.3f\"%OE  )\n",
    "summaryfile = \"%s/%s_%s_enrichment_score.summary.txt\"%(Exportdir, ObsKmerChrom, \"_\".join([str(pb) for pb in ObsKmer]) )\n",
    "with open(summaryfile, \"w\") as fileID:\n",
    "    fileID.writelines( \"\\n\".join(summary_result) )\n",
    "fileID.close()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
